src/analysis/retail/druid/balance/modules/spells/RattleTheStars.tsx

```diff
@@ -0,0 +1,189 @@
+      this.handleDamage,
+    );
+
+    this.addEventListener(Events.fightend, this.handleStacks);
+  }
+
+  handleStacks(
+    event:
+      | ApplyBuffEvent
+      | ApplyBuffStackEvent
+      | RemoveBuffEvent
+      | RemoveBuffStackEvent
+      | FightEndEvent,
+    stack = null,
+  ) {
+    this.buffStacks[this.lastStacks].push(event.timestamp - this.lastUpdate);
+    if (event.type === EventType.FightEnd) {
+      return;
+    }
+    this.lastUpdate = event.timestamp;
+    this.lastStacks = currentStacks(event);
+  }
+
+  handleStarsurgeCast(event: CastEvent) {
+    if (!this.selectedCombatant.hasBuff(SPELLS.TOUCH_THE_COSMOS.id)) {
+      this.totalAPSaved +=
+        (STARSURGE_BASE_COST -
+          (this.elunedGuidanceDiscountActive ? STARSURGE_ELUNES_GUIDANCE_DISCOUNT : 0)) *
+        PERCENT_AP_SAVED_PER_STACK *
+        this.selectedCombatant.getBuffStacks(SPELLS.RATTLED_STARS.id);
+    }
+    this.buffedStarsurges[this.lastStacks] = this.buffedStarsurges[this.lastStacks] + 1;
+  }
+
+  handleStarfallCast(event: CastEvent) {
+    if (!this.selectedCombatant.hasBuff(SPELLS.TOUCH_THE_COSMOS.id)) {
+      this.totalAPSaved +=
+        (STARFALL_BASE_COST -
+          (this.elunedGuidanceDiscountActive ? STARFALL_ELUNES_GUIDANCE_DISCOUNT : 0)) *
+        PERCENT_AP_SAVED_PER_STACK *
+        this.selectedCombatant.getBuffStacks(SPELLS.RATTLED_STARS.id);
+    }
+    this.buffedStarfalls[this.lastStacks] = this.buffedStarfalls[this.lastStacks] + 1;
+  }
+
+  handleDamage(event: DamageEvent) {
+    const buffStacks = this.selectedCombatant.getBuffStacks(SPELLS.RATTLED_STARS.id);
+    this.totalAddedDamage += calculateEffectiveDamage(
+      event,
+      buffStacks * PERCENT_DAMAGE_INCREASE_PER_STACK,
+    );
+  }
+
+  statistic() {
+    const dpsIncrease = this.totalAddedDamage / (this.owner.fightDuration / 1000);
+    return (
+      <Statistic
+        position={STATISTIC_ORDER.CORE(7)}
+        wide
+        size="flexible"
+        tooltip={`This approximate DPS increase of ${formatNumber(
+          dpsIncrease,
+        )} on considers the damage increase from the talent, not increased Starsurge or Starfall usages from astral power that was saved`}
+        dropdown={
+          <>
+            <table className="table table-condensed">
+              <thead>
+                <tr>
+                  <th>Stacks</th>
+                  <th>Starsurges</th>
+                  <th>Starfalls</th>
+                  <th>Uptime</th>
+                </tr>
+              </thead>
+              <tbody>
+                {this.buffStacks.map((e, i) => (
+                  <tr key={i}>
+                    <th>{i} Stacks</th>
+                    <td>{this.buffedStarsurges[i]}</td>
+                    <td>{this.buffedStarfalls[i]}</td>
+                    <td>
+                      {formatPercentage(e.reduce((a, b) => a + b, 0) / this.owner.fightDuration)}%
+                    </td>
+                  </tr>
+                ))}
```

Comment:
i feel like we have something that covers this use-case already but idk what it is 🤔

Reply:
I think I stole this calculation from the Starlord module, would be happy to try something else if we have something better.

Reply:
This should be fine for now - I can get this in